src/IO/Observer/ReductionActions.hpp

```diff
@@ -0,0 +1,266 @@
+          const std::unordered_set<ArrayComponentId>&
+              reduction_component_ids) mutable noexcept {
+          auto& contribute_count =
+              (*reduction_observers_contributed)[observation_id];
+          if (reduction_data_map->count(observation_id) == 0) {
+            reduction_data_map->emplace(observation_id,
+                                        std::move(reduction_data));
+            reduction_names_map->emplace(observation_id, reduction_names);
+            contribute_count = 1;
+          } else {
+            ASSERT(
+                reduction_names_map->at(observation_id) == reduction_names,
+                "Reduction names differ at ObservationId "
+                    << observation_id
+                    << " with the expected names being "
+                    // Use MakeString to get around ADL for STL stream operators
+                    // (MakeString is in global namespace).
+                    << (MakeString{} << reduction_names_map->at(observation_id)
+                                     << " and the received names being "
+                                     << reduction_names));
+            reduction_data_map->operator[](observation_id)
+                .combine(std::move(reduction_data));
+            contribute_count++;
+          }
+
+          // Check if we have received all reduction data from the registered
+          // elements. If so, we reduce to the local ObserverWriter nodegroup.
+          if (contribute_count == reduction_component_ids.size()) {
+            const auto node_id = Parallel::my_node();
+            auto& local_writer = *Parallel::get_parallel_component<
+                                      ObserverWriter<Metavariables>>(cache)
+                                      .ckLocalBranch();
+            Parallel::threaded_action<ThreadedActions::WriteReductionData>(
+                local_writer, observation_id,
+                node_id == 0 ? std::move((*reduction_names_map)[observation_id])
+                             : std::vector<std::string>{},
+                std::move((*reduction_data_map)[observation_id]));
+            reduction_data_map->erase(observation_id);
+            reduction_names_map->erase(observation_id);
+            reduction_observers_contributed->erase(observation_id);
+          }
+        },
+        db::get<Tags::ReductionArrayComponentIds>(box));
+  }
+};
+}  // namespace Actions
+
+namespace ThreadedActions {
+/*!
+ * \ingroup ObserverGroup
+ * \brief Collect the reduction data from the Observer group on the
+ * ObserverWriter nodegroup before sending to node 0 for writing to disk.
+ *
+ * \note This action is also used for writing on node 0.
+ */
+struct WriteReductionData {
+ private:
+  static constexpr void write_data(std::tuple<>&& /*data*/,
+                                   const std::string& /*file_prefix*/,
+                                   std::index_sequence<> /*meta*/) noexcept {}
```

Comment:
This overload is unused.

Reply:
Okay, I changed it to what my intention was: a no-op in the case of empty tuple. Does that sound reasonable?

Reply:
Yes, that makes sense.  Do we want a test of this case?  I can't tell how easy that is to do.

Reply:
I'll give it a shot. It may be as simple as adding a bool template parameter to the `Metavariables` and adding a nearly identical test to what is there, in which case it's definitely worth it :) Otherwise, I think because we'll need to change all of this once we have some idea of performance, getting it to work at the `only these blocks` level, and also with events and triggers I don't expect this code to survive beyond a few months.

Reply:
Okay, I ended up removing the overload and adding a static assert that at least one thing must be reduced. I also added the static assert to `Parallel::ReductionData` since that fails to compile (too many constructors that look like 'ReductionData()`). I think it's reasonable to have people reduce data during a reduction since otherwise it seems like a lot of network activity of empty messages for no reason. How does that sound?

Reply:
That sounds like a good restriction, yeah.